src/pykeen/optimizers/rsgd.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+
+"""Riemannian SGD optimizer.
+
+Code originally written by Ivana Balažević
+https://github.com/ibalazevic/multirelational-poincare/blob/master/rsgd.py
+and used under the MIT License.
+"""
+from typing import Sequence, Set
+
+import torch
+from torch.nn import Parameter
+from torch.optim.optimizer import Optimizer, required
+
+from ..utils import full_p_exp_map
+
+__all__ = [
+    'RiemannianSGD',
+]
+
+
+class RiemannianSGD(Optimizer):
+    """A variant of :class:`torch.optim.SGD` generalized for riemannian manifolds."""
+
+    def __init__(self, params: Sequence[Parameter], param_names: Sequence[str], poincare: Set[str], lr=required):
+        defaults = dict(lr=lr)
+        super().__init__(params, defaults)
+        self._pykeen_extras = {
+            'param_names': param_names,
+            'poincare': poincare,
+        }
+
+    def is_poincare(self, param) -> bool:
+        """Check if the param is a poincare param."""
+        return self._pykeen_extras['param_names'][id(param)] in self._pykeen_extras['poincare']
+
+    def step(self, lr=None):  # noqa:D102
+        loss = None
+        for group in self.param_groups:
+            for i, p in enumerate(group["params"]):
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
+                if lr is None:
+                    lr = group["lr"]
+
+                if self.is_poincare(p):
+                    d_p = poincare_grad(p, d_p)
+                    p.data = poincare_update(p, d_p, lr)
+                else:
+                    p.data = euclidean_update(p, d_p, lr)
+        return loss
+
+
+def euclidean_update(p, d_p, lr):
+    p.data = p.data - lr * d_p
+    return p.data
+
+
+def poincare_grad(p, d_p):
+    p_sqnorm = torch.clamp(torch.sum(p.data ** 2, dim=-1, keepdim=True), 0, 1 - 1e-5)
```

Comment:
The `1 - 1.0e-05` is part of the [Poincare disk model](https://en.wikipedia.org/wiki/Poincar%C3%A9_disk_model), right? I am not very deep into the topic, but is RiemannianSGD restricted to this model? At least when skimming [this text](https://andbloch.github.io/Stochastic-Gradient-Descent-on-Riemannian-Manifolds/) it seemed like the method is more general.

So maybe we should document this limitation(?) somewhere.

btw, the last mentioned blog post refers to a library https://github.com/geoopt/geoopt which seems to implement RiemannianSGD along other optimizers in PyTorch.